lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp

```diff
@@ -672,6 +689,240 @@ TEST_F(SessionManagerHandlerTest, test_SetAmfSessionAmbr) {
+      &server_context, &request,
+      [this](grpc::Status status, SmContextVoid Void) {});
+
+  // Run session creation in the EventBase loop
+  evb->loopOnce();
+  evb->loopOnce();
+  SessionConfig cfg;
+  cfg.common_context       = request.common_context();
+  cfg.rat_specific_context = request.rat_specific_context();
+  cfg.rat_specific_context.mutable_m5gsm_session_context()->set_ssc_mode(
+      SSC_MODE_3);
+
+  auto session_map = session_store->read_sessions({IMSI1});
+  auto it          = session_map.find(IMSI1);
+  EXPECT_FALSE(it == session_map.end());
+  EXPECT_EQ(session_map[IMSI1].size(), 1);
+  auto& session_temp = session_map[IMSI1][0];
+  EXPECT_EQ(session_temp->get_config().common_context.sid().id(), IMSI1);
+  SessionUpdate update = SessionStore::get_default_session_update(session_map);
+  uint32_t pdu_id      = 5;
+  SessionSearchCriteria id1_success_sid(IMSI1, IMSI_AND_PDUID, pdu_id);
+  auto session_it = session_store->find_session(session_map, id1_success_sid);
+  auto& session   = **session_it;
+  auto session_i  = session->get_session_id();
+  EXPECT_CALL(*events_reporter, session_terminated(IMSI1, testing::_)).Times(1);
+  session_enforcer->m5g_complete_termination(
+      session_map, IMSI1, session_i, update);
+
+  EXPECT_EQ(session_map[IMSI1].size(), 0);
+}
+
+TEST_F(SessionManagerHandlerTest, test_SessionCreationFailure_EventReport) {
+  magma::SetSMSessionContext request;
+  auto* req =
+      request.mutable_rat_specific_context()->mutable_m5gsm_session_context();
+  auto* reqcmn = request.mutable_common_context();
+  req->set_pdu_session_id({0x5});
+  req->set_request_type(magma::RequestType::INITIAL_REQUEST);
+  req->mutable_pdu_address()->set_redirect_address_type(
+      magma::RedirectServer::IPV4);
+  req->mutable_pdu_address()->set_redirect_server_address("10.20.30.40");
+  req->set_priority_access(magma::priorityaccess::High);
+  req->set_imei("123456789012345");
+  req->set_gpsi("9876543210");
+  req->set_pcf_id("1357924680123456");
+
+  reqcmn->mutable_sid()->set_id("IMSI000000000000001");
+  reqcmn->set_sm_session_state(magma::SMSessionFSMState::CREATING_0);
+  reqcmn->set_rat_type(magma::RATType::TGPP_LTE);
+
+  grpc::ServerContext server_context;
+  SessionConfig cfg;
+  cfg.common_context       = request.common_context();
+  cfg.rat_specific_context = request.rat_specific_context();
+  cfg.rat_specific_context.mutable_m5gsm_session_context()->set_ssc_mode(
+      SSC_MODE_3);
+
+  EXPECT_CALL(*events_reporter, session_create_failure(cfg, testing::_))
+      .Times(1);
+
+  set_session_manager->SetAmfSessionContext(
+      &server_context, &request,
+      [this](grpc::Status status, SmContextVoid Void) {});
+
+  // Run session creation in the EventBase loop
+  evb->loopOnce();
+  evb->loopOnce();
+
+  auto session_map = session_store->read_sessions({IMSI1});
+  auto it          = session_map.find(IMSI1);
+  EXPECT_FALSE(it == session_map.end());
+  EXPECT_EQ(session_map[IMSI1].size(), 0);
+}
```

Comment:
As mentioned our unit test are not per function but per process.
You can collapse 
```
test_create_session_policy_report
test_SessionCreation_EventReporter
test_terminate_session_policy_report
test_SessionTerminate_EventReport
```
in one single test and reduce repetition. 

Also I don't see this code testing what you changed on the codebase. You should be creating/modifiying [test_local_enforcer](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Amagma%2Fmagma.git&branch=master&file=lte%2Fgateway%2Fc%2Fsession_manager%2Ftest%2Ftest_local_enforcer.cpp&editor=JetBrains&version=v1.2.1&utm_product_name=CLion&utm_product_version=2021.1.2&start_row=366&start_col=3&end_row=366&end_col=3)

You should be adding test to any of our test files taht cover any of the changes you did. For example, since you have changed `session_state` it would be good to add or modify a test on test_session_State.  A way to find what test to modify is to look for the function you have modified and check if they exist in any test . For example, see how aggregate_recods that you modified is all around the place 

![Uploading image.png…]()

Reply:
I have clubbed `test_create_session_policy_report and test_SessionCreation_EventReporter` in one test case for session creation and `test_terminate_session_policy_report and test_SessionTerminate_EventReport` for session termission.

I have added few more UT for creating/modifying usage report with `aggregate_records` for 5g.

It is difficult to cover of UT as same level of 4g with pressure of deliverables of many features in 5G. But subsequently we will add more UT for corner cases.

Reply:
i will review this more in depth but let's fix first that function that seems to have a bug

Reply:
Yes, I have added the comments in respective places and made separate sections for understanding wise.